bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/sitemap/ProductsSitemapGenerator.java

```diff
@@ -0,0 +1,167 @@
+        }
+
+        int currentIndex = context.getProperty(PN_NEXT_PRODUCT, 0);
+        int currentPageIndex = context.getProperty(PN_NEXT_PAGE, 1);
+        int maxPages = Integer.MAX_VALUE;
+        // parameter map to be reused while iterating
+        UrlProvider.ParamsBuilder paramsBuilder = new UrlProvider.ParamsBuilder()
+            .page(externalizer.externalize(sitemapRoot));
+
+        while (currentPageIndex <= maxPages) {
+            String query = Operations.query(productsQueryFor(currentPageIndex, pageSize)).toString();
+            GraphqlResponse<Query, Error> resp = graphql.execute(query);
+
+            if (resp.getErrors() != null && resp.getErrors().size() > 0) {
+                SitemapException ex = new SitemapException("Failed to execute graphql query.");
+                resp.getErrors().forEach(error -> ex.addSuppressed(new Exception(error.getMessage())));
+                throw ex;
+            }
+
+            Products products = resp.getData().getProducts();
+            List<ProductInterface> items = products.getItems();
+            maxPages = products.getTotalCount() / pageSize;
+
+            if (products.getTotalCount() % pageSize > 0) {
+                // there is a fractional part of items on the last page
+                maxPages++;
+            }
+
+            for (int i = currentIndex; i < items.size(); i++) {
+                ProductInterface product = items.get(i);
+                if (productFilter != null && !productFilter.shouldInclude(productPage, product)) {
+                    LOGGER.debug("Ignore product {}, not allowed by filter: {}", product.getSku(), productFilter.getClass()
+                        .getSimpleName());
+                    continue;
+                }
+                Map<String, String> params = paramsBuilder
+                    .sku(product.getSku()).urlKey(product.getUrlKey())
+                    .variantSku(null).variantUrlKey(null)
+                    .map();
+                sitemap.addUrl(urlProvider.toProductUrl(null, null, params));
+                context.setProperty(PN_NEXT_PRODUCT, i + 1);
+            }
+
+            currentIndex = 0;
+            context.setProperty(PN_NEXT_PRODUCT, currentIndex);
+            context.setProperty(PN_NEXT_PAGE, ++currentPageIndex);
+        }
+    }
+
+    private static QueryQueryDefinition productsQueryFor(int pageIndex, int pageSize) {
+        return q -> q.products(
+            arguments -> arguments
+                .search(StringUtils.EMPTY)
+                .pageSize(pageSize)
+                .currentPage(pageIndex),
+            resultSet -> resultSet
+                .totalCount()
+                .items(product -> product
+                    .urlKey()
+                    .sku()
+                    .onConfigurableProduct(configurableProduct -> configurableProduct
+                        .variants(variant -> variant.product(SimpleProductQuery::sku)))));
```

Comment:
When is this used by the code? In L141 the variant params are always null.

Reply:
Good catch